src/Kernel/RectorKernel.php

```diff
@@ -101,4 +119,82 @@ private function createDefaultConfigFiles(): array
     {
         return [__DIR__ . '/../../config/config.php'];
     }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function createConfigsHash(array $configFiles): string
+    {
+        $fileHasher = new Hasher();
+
+        if (self::$defaultFilesHash === null) {
+            self::$defaultFilesHash = $fileHasher->hashFiles($this->createDefaultConfigFiles());
+        }
+
+        Assert::allString($configFiles);
+        $configHash = $fileHasher->hashFiles($configFiles);
+
+        return self::$defaultFilesHash . $configHash;
+    }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function buildContainer(array $configFiles, string $hash = null): ContainerBuilder
+    {
+        $defaultConfigFiles = $this->createDefaultConfigFiles();
+        $configFiles = array_merge($defaultConfigFiles, $configFiles);
+
+        $compilerPasses = $this->createCompilerPasses();
+
+        $configureCallMergingLoaderFactory = new ConfigureCallMergingLoaderFactory($this->configureCallValuesCollector);
+        $containerBuilderFactory = new ContainerBuilderFactory($configureCallMergingLoaderFactory);
+
+        $containerBuilder = $containerBuilderFactory->create($configFiles, $compilerPasses);
+
+        // @see https://symfony.com/blog/new-in-symfony-4-4-dependency-injection-improvements-part-1
+        $containerBuilder->setParameter('container.dumper.inline_factories', true);
+        // to fix reincluding files again
+        $containerBuilder->setParameter('container.dumper.inline_class_loader', false);
+
+        $containerBuilder->compile();
+
+        return $this->container = $containerBuilder;
+    }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function buildCachedContainer(array $configFiles): ContainerInterface {
+        $hash = $this->createConfigsHash($configFiles);
+
+        $filesystem = new SmartFileSystem();
+        $className = 'RectorKernel'.$hash;
+        $file = sys_get_temp_dir() .'/rector/'. self::CACHE_KEY .'-'.$hash.'.php';
+
+        if (file_exists($file)) {
+            require_once $file;
+            $className = '\\'.__NAMESPACE__ .'\\'. $className;
+            $container = new $className();
+            if (!$container instanceof ContainerInterface) {
+                throw new ShouldNotHappenException();
+            }
+        } else {
+            $container = $this->buildContainer($configFiles, $hash);
+
+            $dumper = new PhpDumper($container);
+            $dumpedContainer = $dumper->dump([
+                'class' => $className,
+                'namespace' => __NAMESPACE__
+            ]);
+            if (!is_string($dumpedContainer)) {
+                throw new ShouldNotHappenException();
+            }
+
+            $filesystem->dumpFile($file, $dumpedContainer);
+        }
+
+        return $container;
+    }
+
 }
```

Comment:
This class now got a bit complex. Could you cover those new methods with simple test, so we're ready for any breaks?

Thanks :+1:

Reply:
tbh I am not sure what to test. only implementation details of the kernel changed.

Reply:
Ideally make this a public method and test the expected result has specific attributes.

Reply:
I broke up the logik into smaller classes. wdyt?

Reply:
Much better, thank you :+1: